### PR TITLE
Convert GPX/KML to geojson

### PR DESCRIPTION
**Pull request by GretaCB**
Per https://github.com/mapbox/preprocessorcerer/issues/19

- [x] Convert each layer of KML/GPX into geojson file (max 15 layers)
- [x] Add index file for each new file/layer > 10MB
- [x] [Validate layer names](https://github.com/mapbox/preprocessorcerer/commit/1155713d6147615843fe68a2d765dd2232fbed03#commitcomment-16163231)
- [x] Include metadata file for original KML/GPX (run through mapnik-omnivore)
- [x] Drop GPX `track_points`. Found that these are automatically inserted by gdal (most of the time). @BergWerkGIS did we decide on if possible to allow `track_points` produced by an actual GPS device?
- [x] Tests (anything we're not testing that we should?)

@BergWerkGIS

### CODE REVIEW

**Comment by GretaCB** [comment]
Per https://github.com/mapbox/unpacker/issues/907, just needs to run mapnik-omnivore and add the output as a JSON file to the bundle (for both KML/GPX preprocessor and geojson-index preprocessor). Happy to jump on this tomorrow.

**Comment by GretaCB** [comment]
The togeojson KML tests fail 50% of the time locally with the following error:
```
not ok 9 .end() called twice
  ---
    operator: fail
  ...

/Users/mapbox/dev/preprocessorcerer/node_modules/queue-async/build/queue.js:41
        if (!tasks[i]) throw new Error; // detect multiple callbacks
```

Going to split each fixture into its own tape test as a first step. Wonder if a previous test in the for loop is erroring and an assert.end() is lingering.

**Comment by GretaCB** [comment]
Looks like the original error that was causing the .end() callback error is:
```
max tree depth:8
split ratio:0.55
processing '/var/folders/2p/m7jcvfp90wngshgr9j0cd9tc0000gn/T/f32881494358414c/empty-features-only.geojson' as GeoJSON
Mapnik LOG> 2016-03-18 14:39:26: Error loading mapped memory file: '/var/folders/2p/m7jcvfp90wngshgr9j0cd9tc0000gn/T/f32881494358414c/empty-features-only.geojson' (Invalid argument)
Error : cannot memory map /var/folders/2p/m7jcvfp90wngshgr9j0cd9tc0000gn/T/f32881494358414c/empty-features-only.geojson
Error: failed to process /var/folders/2p/m7jcvfp90wngshgr9j0cd9tc0000gn/T/f32881494358414c/empty-features-only.geojson
done!
```

I suspect mapnik doesnt like some of the geojson files we are feeding it for indexing or some kind of race condition when indexing each layer?

cc @artemp @springmeyer

**Comment by GretaCB** [comment]
Per chat with @springmeyer , looks like the indexing error is pointing to something else going on in the converted geojson layers.

I took a look at the resulting converted geojson layers for the tests that are failing, and looks as if they are  producing [incomplete](https://gist.github.com/GretaCB/3a257dd213a1df4a2c6a) [objects](https://gist.github.com/GretaCB/dd7ba2776580d46936f0).

Going to keep :eyes:

**Comment by BergWerkGIS** [comment]
One step further with debugging the [failing test](https://github.com/mapbox/preprocessorcerer/blob/43d2630ddaf4b94ddaf969fb7458f4889fd1d4ea/test/index.test.js#L18-L31).

I have proof now, that something is still holding on to `tmpfile` when it should be deleted:

image

Unfortunately it's a pain to find out: running the test from the console the exception occurs every time.
Slowly stepping thru the test in a debug session works.
Jumping between different break points in a debug session fails.
I've already tried different things (without success):
* delay before delete
* nulling `src` in wmtiff
* ...

**Comment by GretaCB** [comment]
I think we have here a good ole fashioned control flow problem (when I remove createIndex locally, tests no longer fail). Looks like createIndex can potentially trigger before the forEach loops are finished. This is because createIndex is being called outside of the forEach loops, and is not relying on any sort of callback or way to determine when layers are in fact ready to be indexed.

I like your idea from yesterday, that we should index _after_ all layers are converted. We can also refactor the main function a bit to better manage flow.

Found a [great example](http://howtonode.org/control-flow) of how we can try untangling some of this:
``` js
var fs = require('fs');

function read_directory(path, next) {
  fs.readdir(".", function (err, files) {
    var count = files.length,
        results = {};
    files.forEach(function (filename) {
      fs.readFile(filename, function (data) {
        results[filename] = data;
        count--;
        if (count <= 0) {
          next(results);
        }
      });
    });
  });
}

function read_directories(paths, next) {
  var count = paths.length,
      data = {};
  paths.forEach(function (path) {
    read_directory(path, function (results) {
      data[path] = results;
      count--;
      if (count <= 0) {
        next(data);
      }
    });
  });
}

read_directories(['articles', 'authors', 'skin'], function (data) {
  // Do something
});
```

**Comment by BergWerkGIS** [comment]
Nothing wrong with the control flow, just gdal has to be manually disposed to get rid of the file reference faster:
* preprocessorcerer: https://github.com/mapbox/preprocessorcerer/commit/8dc42912a9deff3d17ac64bdfba3f4ecb6cac4df
* wmtiff: https://github.com/mapbox/node-wmtiff/commit/6c809733b31627e3afdb4e4921c5238a0a2ff6ba

The 4 references to the dataset in the [screenshot](https://github.com/mapbox/preprocessorcerer/pull/61#issuecomment-199324218) above came from
* wmtiff
* tif-reproject.preprocessor.js
* tif-toBytes.preprocessor.js
* preprocessor.criteria

as each of them opens the dataset (maybe room for improvement?) and `ds` is still holding on to the file (despite of `ds.close()`).
Maybe because everything is done in callbacks and node hasn't yet let go of the objects.

----

Working on the next failing test now:

```
# [parts mbtiles] correct number of parts for mbtiles without grids table
ok 40 no error
ok 41 four parts

fs.js:772
  return binding.unlink(pathModule._makeLong(path));
                 ^
Error: EBUSY, resource busy or locked 'C:\Users\ADMINI~1\AppData\Local\Temp\2\79c7b0a1613d996b'
    at Object.fs.unlinkSync (fs.js:772:18)
    at c:\mb\preprocessorcerer\test\mbtiles-byTiles.test.js:17:10
    at Statement.<anonymous> (c:\mb\preprocessorcerer\parts\mbtiles-byTiles.js:25:9)
npm ERR! Test failed.  See above for more details.
```

**Comment by BergWerkGIS** [comment]
Same with failing mbtiles test: resources are not properly disposed. 
In this case connection to the database is not closed.
Fixed: https://github.com/mapbox/preprocessorcerer/commit/abf63bcb9b80b9580a9c4a07eee566f6c8ce33f5

-----

Now all tests run through again, but 8 are failing.
Continuing the test quest ... :smirk:

**Comment by BergWerkGIS** [comment]
> I think we have here a good ole fashioned control flow problem

Yes, you are right. This is a control flow problem.

I didn't realize that you were talking about [GPX test](https://github.com/mapbox/preprocessorcerer/blob/abf63bcb9b80b9580a9c4a07eee566f6c8ce33f5/test/togeojson-gpx.test.js#L47-L57) and I was talking about [tiff test](https://github.com/mapbox/preprocessorcerer/blob/abf63bcb9b80b9580a9c4a07eee566f6c8ce33f5/test/index.test.js#L10-L29)

[Here](https://github.com/mapbox/preprocessorcerer/blob/togeojson/preprocessors/togeojson-gpx.preprocessor.js#L85-L87) `createIndex` is called async, but at the same time control flow continues to [here](https://github.com/mapbox/preprocessorcerer/blob/togeojson/preprocessors/togeojson-gpx.preprocessor.js#L90-L102)

**Comment by BergWerkGIS** [comment]
That's also why `.end()` is called twiced: 1x from `createIndex()` and 1x from the exported function itself.

<pre>
not ok 4 expected error message
  ---
    operator: equal
    expected: 'GPX does not contain any valid features.'
    actual:   undefined
  ...
</b>not ok 5 .end() called twice</b>
  ---
    operator: fail
  ...
</pre>

**Comment by BergWerkGIS** [comment]
`npm test` works for me now - no failed tests.
But I have to use https://github.com/mapbox/node-wmtiff/commit/6c809733b31627e3afdb4e4921c5238a0a2ff6ba.

**Comment by GretaCB** [comment]
Node `v0.10.40` worked fine on OSX as well. I took a look around preprocessorcer and did a little research, but didnt find much that explains the immortal tmpfile. I did fine [this](https://github.com/naturalatlas/node-gdal/issues/93#issuecomment-84084696)

> Datasets get closed automatically when node garbage collects them, but there is no guarantee that node will garbage collect before exiting.

This lead me down a google search of node GC, but not sure that's the proper route at this point. It's odd that the tmpfile still exists even when running fs.unlink. Will keep pondering.

**Comment by BergWerkGIS** [comment]
> I did find [this](https://github.com/naturalatlas/node-gdal/issues/93#issuecomment-84084696)

Great find. Thanks.
This was exactly my hunch:
**_"Making sure the dataset is closed and flushed before the process exits fixes the problem for me. Datasets get closed automatically when node garbage collects them, but there is no guarantee that node will garbage collect before exiting."_**

`but there is no guarantee that node will garbage collect before exiting` or in our case before we try to delete the dataset.

> but didnt find much that explains the immortal tmpfile

The explanation above was the reason for not being able to delete the tmpfile, as node hat not yet done its garbage collection.

Sorry if I caused any confusion, but this should be fixed now and the file should not be immortal anymore.

`ds.close()` and `ds = null` I've put into the code at various places should take care of disposing the objects (references to the file).

> This lead me down a google search of node GC, but not sure that's the proper route at this point. Will keep pondering.

Not necessary to spend more time on this.
I've literally spent weeks on node GC when investigating strange `node-mapnik` crashes some time ago.

There are GC parameters you can pass to the node executable to fiddle with the internals of its GC
e.g.: `--expose-gc`, or others to change memory usage before GC kicks in, or to prevent GC at all.
But none of them really helped and we decided to stick with the defaults.

-------

The take away here is: _always dispose of resources properly as soon as you are done with them._

Be it a file (`fs.close()`), a database connection (`db.close()`) or another native wrapper (like gdal).

**Comment by springmeyer** [comment]
> The take away here is: always dispose of resources properly as soon as you are done with them.

Awesome work. Also agree @BergWerkGIS. This is a pretty classic case of windows behavior being more strict and that helping highlight core issues of the code that should be written more robustly. :100: 

> ds.close() and ds = null I've put into the code at various places should take care of disposing the objects (references to the file).

That is neat/great that setting to `null` triggers the actual close (I presume because it triggers the garbage collection sooner). I'm left however thinking that it is probably worth bringing this up again with Brian and Brandon. I'm sure the current `close()` behavior of not releasing until garbage collection makes sense given the design and object inheritance in node-gdal. But nevertheless it seems like odd behavior. Perhaps they would reconsider having `ds.close()` actually close the underlying gdal file handle directly or perhaps adding a `ds.closeNow()` method to force it. I've not looked at the code enough to know if this is feasible, but I think it would be great @BergWerkGIS  if you would raise this for them in a ticket to flesh this out further.

**Comment by BergWerkGIS** [comment]
@springmeyer Already looked at the code and I think this is purely on node's side (and our implementation) and there is nothing else gdal can do about it (although not familiar with node's GC internals).

[`ds.close()`](https://github.com/naturalatlas/node-gdal/blob/e9dfb55a9699593483c4e4d836e805200b36c3d6/src/gdal_dataset.cpp#L307-L319) calls [`dispose`](https://github.com/naturalatlas/node-gdal/blob/e9dfb55a9699593483c4e4d836e805200b36c3d6/src/gdal_dataset.cpp#L91-L116) to null the internal resources:

```cpp
	if (this_dataset) {
		LOG("Disposing Dataset [%p]", this_dataset);
		ptr_manager.dispose(uid);
		LOG("Disposed Dataset [%p]", this_dataset);
		this_dataset = NULL;
	}
```

----

Pseudo code of our implementation looks like this:

```js
function one(){var ds = new gdal.ds; return two(ds);}
function two(ds){return three(ds);}
function three(ds){return four(ds);}
function four(){fs.unlink(ds);}
```

And with this implementation it absolutely makes sense for me that node hasn't fully released `ds` yet, as in method `four()` we are essentially still in method `one()`.
Or am I missing something about node/js callbacks?

**Comment by GretaCB** [comment]
It happened to me after adding a few more `fs.unlink`s to clean up a bit in other tests:
```
# [spatial-index] indexes (input folder output file)
ok 108 should be equal
ok 109 no error
ok 110 (unnamed assert)

fs.js:772
  return binding.unlink(pathModule._makeLong(path));
                 ^
Error: EPERM, operation not permitted '/var/folders/2p/m7jcvfp90wngshgr9j0cd9tc0000gn/T/0fa9a8e51db16a2e'
    at Object.fs.unlinkSync (fs.js:772:18)
    at /Users/mapbox/dev/preprocessorcerer/test/spatial-index.test.js:64:12
    at ReadStream.<anonymous> (/Users/mapbox/dev/preprocessorcerer/node_modules/checksum/checksum.js:84:9)
    at ReadStream.emit (events.js:117:20)
    at _stream_readable.js:944:16
    at process._tickCallback (node.js:448:13)
```

**Comment by BergWerkGIS** [comment]
Interesting, doesn't happen for me, neither with
```
node node_modules\tape\bin\tape test\spatial-index.test.js
```
nor with
```
npm test
```

@GretaCB could you try with this commit?  https://github.com/mapbox/preprocessorcerer/pull/61/commits/a08b2cac71947f60592f342a57a65114936a54c9

**Comment by GretaCB** [comment]
Ok, looks like the funky travis test is due to multiple node-mapnik versions:

```
├── mapnik@3.5.2 
└─┬ mapnik-omnivore@7.1.1
  └── mapnik@3.4.17 
```

Per https://github.com/mapbox/mapnik-internal/issues/73#issuecomment-200550995, we're going to be pushing unpacker to node-mapnik 3.5.0, so going to update to mapnik-omnivore `7.3.0` (which contains the newest node-mapnik)

**Comment by GretaCB** [comment]
UPDATE on funky test failure

[Per chat with Jake and Blake](https://mapbox.slack.com/archives/mapnik/p1458847939004265), this is a build bug in node-mapnik and will be fixed in the newest push.

**Comment by BergWerkGIS** [comment]
Cannot reproduce the `unlink` problem locally also Travis didn't have a problem on the last run:
https://travis-ci.org/mapbox/preprocessorcerer/builds/118480437

-----

The highlighted files are those left after `npm test`.
Will take a look where there are coming from.

image

**Comment by GretaCB** [comment]
Hmmm, looking into why tests are still failing here, so we can get this ready to rock.

File: preprocessors/spatial-index.preprocessor.js [hunk 1]
```diff
@@ -34,6 +34,7 @@ module.exports = function(infile, outdir, callback) {
         .on('exit', function() {
           // If error printed to --validate-features log
           if (data.indexOf('Error') != -1) {
+            console.log(data);
```
Comment (rclark) [inline preprocessors/spatial-index.preprocessor.js]: debug logging?

Comment (GretaCB) [inline preprocessors/spatial-index.preprocessor.js]: 👍  thank you

File: preprocessors/togeojson-gpx.preprocessor.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    try {
+      ds_gpx = gdal.open(infile);
+    }
+    catch (err) {
+      return callback(new Error(err));
+    }
+
+    ds_gpx.layers.forEach(function(lyr_gpx) {
+      //drop point layers as they can get really huge
+      if (lyr_gpx.name === 'track_points' || lyr_gpx.name === 'route_points') {
```
Comment (rclark) [inline preprocessors/togeojson-gpx.preprocessor.js]: Does this represent a breaking change from present behavior? That is, would these layers be included in a GPX upload today?

File: preprocessors/spatial-index.preprocessor.js (hunk 1 above)
Comment (BergWerkGIS) [inline preprocessors/spatial-index.preprocessor.js]: @GretaCB what's the plan of attack?
This has changed here (https://github.com/mapbox/preprocessorcerer/compare/stderr-fix#diff-39694ce4a586caad3e6ad813c295b454R32) too as latest `mapnik-index` is now returning an error code, if `--validate-features` fails.
Need to wait to update to latest node-mapnik?

File: preprocessors/togeojson-gpx.preprocessor.js (hunk 2 above)
Comment (GretaCB) [inline preprocessors/togeojson-gpx.preprocessor.js]: Yes, they would. See discussion per https://github.com/mapbox/unpacker/issues/871#issuecomment-197206459

File: preprocessors/spatial-index.preprocessor.js (hunk 1 above)
Comment (GretaCB) [inline preprocessors/spatial-index.preprocessor.js]: Plan of attack re: https://github.com/mapbox/mapnik-swoop/pull/14
cc @mapsam

Comment (mapsam) [inline preprocessors/spatial-index.preprocessor.js]: Working through a node-mapnik dev build so we can test this right now.

**Pull request merged**